packages/flame_studio/lib/src/core/component_tree.dart

```diff
@@ -0,0 +1,81 @@
+import 'package:flame/components.dart';
+import 'package:flame/game.dart';
+import 'package:flame_studio/src/core/game_controller.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+final componentTreeProvider =
+    StateNotifierProvider<ComponentTreeObserver, ComponentTreeState>((ref) {
+  final gameState = ref.watch(gameControllerProvider);
+  return ComponentTreeObserver(gameState.game as FlameGame?);
+});
+
+@immutable
+class ComponentTreeNode {
+  const ComponentTreeNode._(this.component, this.children);
+
+  factory ComponentTreeNode.fromComponent(Component component) {
+    return ComponentTreeNode._(
+      component,
+      component.hasChildren
+          ? component.children.map(ComponentTreeNode.fromComponent).toList()
+          : null,
+    );
+  }
+
+  final Component component;
+  final List<ComponentTreeNode>? children;
+  String get name => component.runtimeType.toString();
+  bool get hasChildren => children?.isNotEmpty ?? false;
+
+  @override
+  bool operator ==(Object other) =>
+      other is ComponentTreeNode &&
+      component == other.component &&
+      listEquals(children, other.children);
+
+  @override
+  int get hashCode => Object.hash(component, children);
+}
+
+@immutable
+class ComponentTreeState {
+  ComponentTreeState(Component? rootComponent)
+      : root = rootComponent == null
+            ? null
+            : ComponentTreeNode.fromComponent(rootComponent);
+
+  final ComponentTreeNode? root;
+
+  @override
+  bool operator ==(Object other) =>
+      other is ComponentTreeState && root == other.root;
+
+  @override
+  int get hashCode => root?.hashCode ?? 0;
+}
+
+class ComponentTreeObserver extends StateNotifier<ComponentTreeState> {
+  ComponentTreeObserver(Component? rootComponent)
+      : super(ComponentTreeState(rootComponent)) {
+    if (rootComponent != null) {
+      _refresh(Duration.zero);
+    }
+  }
+
+  static Duration refreshFrequency = const Duration(milliseconds: 300);
```

Comment:
```suggestion
  static const Duration refreshFrequency = Duration(milliseconds: 300);
```

Reply:
This parameter is intended to be modifiable by the user, possibly even within the UI once we have a settings panel.

Reply:
This should probably be handled by a provider instead of a static variable if it should be modifiable.

Reply:
Not really, since nothing needs to be rebuilt when this value changes. The new value will be used automatically on the next update tick.

Reply:
we could make it const for now and refactor once it is possible to change?

Reply:
It's already possible to change, only via code, not UI.